File: linked_list/double_ll.py
# Full implementation of a double linked list DSA in Python

# Node class
class Node:
    # Constructor
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

# Linked list class


class LinkedList:
    # Constructor
    def __init__(self):
        self.head = None

    # Methods
    # Insert first in the list
    def insert_first(self, data):
        # Create a new node
        newNode = Node(data)
        # Point the next link to the head
        newNode.next = self.head
        # Point the previous link to NULL
        newNode.prev = None
        # Check if the head is null, readjust the head with a new node
        if self.head == None:
            self.head = newNode
            newNode.prev = None
        else:
            self.head.prev = newNode
            self.head = newNode

    # Insert after
    def insert_after(self, prev_node, data):
        # Check if the previous node is null
        if not prev_node:
            print('The previous node cannot be null!')
            return
        # Create a new node
        newNode = Node(data)
        # Point the next link of the new node to the next link of the previous node
        newNode.next = prev_node.next
        # Point the previous link of the new node to the previous node
        newNode.prev = prev_node
        # Point the next link of the previous node to the new node to readjust links
        prev_node.next = newNode
        # Check if the new node is not the last one, connect the last node to the new node
        if newNode.next:
            newNode.next.prev = newNode

    # Insert last
    def insert_last(self, data):
        # Create a temp variable to traverse the list
        temp = self.head
        # Create a new node
        newNode = Node(data)
        # Point the next link to null since it's going to be the last node in the list
        newNode.next = None
        # Checkf if there's no head, set the new node to the head and return
        if self.head == None:
            self.head = newNode
            self.head.prev = None
        # Traverse the list till the end and point the last node to the new node, then point the new node to the previous one to readjust the links
        while temp.next:
            temp = temp.next
        temp.next = newNode
        newNode.prev = temp

    # Delete node
    def delete_node(self, del_node):
        # Check if there's no head or no del_node
        if not self.head or not del_node:
            print('There\'s nothing to be deleted')
            return
        # Checkf if the del_node is in the head, shift the head to the next node
        if self.head == del_node:
            self.head = del_node.next
        # Checkf if the del_node is not the last node in the list change the pointers to delete the node and readjust the links
        if del_node.next:
            del_node.next.prev = del_node.prev
        # Check if the del_node is not in the head
        if del_node.prev:
            del_node.prev.next = del_node.next

    # Print the double linked list
    def print_list(self):
        # Traverse the list and on each iteration print values of the nodes
        temp = self.head
        while (temp):
            print(f'Linked list: {temp.data} ', end='')
            temp = temp.next
        print()


if __name__ == '__main__':
    # Instantiate the Linked list class
    ll = LinkedList()
    ll.insert_first(25)
    ll.insert_first(30)
    ll.insert_first(45)
    ll.insert_first(11)
    ll.insert_after(ll.head, 110)
    ll.insert_after(ll.head.next.next, 250)
    ll.insert_last(295)
    ll.insert_last(333)
    ll.insert_first(555)
    ll.delete_node(ll.head.next)
    ll.print_list()

File: linked_list/double_ll.cpp
/* Full implementation of a double linked list with C++ in the object-oriented-fashion */
#include <iostream>

/* Node class */
class Node
{
public:
    int m_data;
    Node *next;
    Node *prev;
    // Default constructor
    Node() = default;
    // Parameterized constructor
    Node(int data)
        : m_data(data), next(NULL), prev(NULL)
    {
    }
};

/* Linked list class */
class LinkedList
{
public:
    Node *head;
    // Default constructor
    LinkedList()
        : head(NULL)
    {
    }

    /* Public member functions */

    // Insert first
    void insertFirst(int data);

    // Insert after
    void insertAfter(Node *prevNode, int data);

    // Insert last
    void insertLast(int data);

    // Delete node
    void deleteNode(Node *delNode);

    // Print linked list
    void printList();
};

void LinkedList::insertFirst(int data)
{
    // Dynamic memory allocation on the heap for a new node
    Node *newNode = new Node();
    // Assign values to the allocated memory (type of Node)
    newNode->m_data = data;
    // Change the next pointer of the new node to the head
    newNode->next = head;
    // Change the previous pointer of the new node to NULL (make it the head node)
    newNode->prev = NULL;
    // Check if the head is NULL
    if (!head)
    {
        head = newNode;
        newNode->prev = NULL;
    }
    else
    {
        head->prev = newNode;
        head = newNode;
    }
}

void LinkedList::insertAfter(Node *prevNode, int data)
{
    // Check if there's no previous node
    if (!prevNode)
    {
        std::cout << "The previous node cannot be null!";
        return;
    }
    // Dynamically memory allocation on the heap for a new node
    Node *newNode = new Node();
    // Assign values
    newNode->m_data = data;
    // Change the next link of the new node to the next link of the previous node
    newNode->next = prevNode->next;
    // Change the previous link of the new node to the previous node
    newNode->prev = prevNode;
    // Change the link of the next node of the previous node to the new node
    prevNode->next = newNode;
    // Check if the new node is not the last node in the list, change the previous pointer of the next node to the new node (to readjust links)
    if (newNode->next)
    {
        newNode->next->prev = newNode;
    }
}

void LinkedList::deleteNode(Node *delNode)
{
    // Check if there's no head or no delNode
    if (!head || !delNode)
    {
        std::cout << "No data to be deleted!";
        return;
    }
    // Check if the delNode is in the head, shift the head node to the next node
    if (head == delNode)
    {
        head = delNode->next;
    }
    // Check if the delNode is not the last node in the linked list
    // Point the previous pointer of the next node to the previous pointer of the current node
    if (delNode->next)
    {
        delNode->next->prev = delNode->prev;
    }
    // Check if the delNode is not in the head
    if (delNode->prev)
    {
        delNode->prev->next = delNode->next;
    }
    // Free memory for delNode pointer
    delete delNode;
}

void LinkedList::insertLast(int data)
{
    // Dynamically memory allocation on the heap for a new node
    Node *newNode = new Node();
    // Set temp variable to traverse the linked list
    Node *temp = head;
    // Assign values
    newNode->m_data = data;
    // Point the next pointer of the new node to NULL since it's going to be the last node
    newNode->next = NULL;
    // Check if the head points to null, change the head to the new node and return
    if (!head)
    {
        head = newNode;
        newNode->prev = NULL;
        return;
    }
    // Traverse the linked list from the head, point the last node to the new node, point the new node to the previous node
    while (temp->next)
    {
        temp = temp->next;
    }
    temp->next = newNode;
    newNode->prev = temp;
}

void LinkedList::printList()
{
    // Traverse the linked list till the end and print values of each node
    Node *temp = head;
    while (temp)
    {
        std::cout << "Linked list: " << temp->m_data << " ";
        temp = temp->next;
    }
    std::cout << std::endl;
}

int main()
{
    // Instantiate the linked list class
    LinkedList ll;
    ll.insertFirst(25);
    ll.insertFirst(30);
    ll.insertFirst(40);
    ll.insertAfter(ll.head->next, 100);
    ll.insertLast(192);
    ll.insertLast(-192);
    ll.insertAfter(ll.head->next->next->next->next, 302);
    ll.insertFirst(35);
    ll.printList();
    return 0;
}